src/ProxyProtocol.cc

```diff
@@ -0,0 +1,306 @@
+    if ((command & 0xFE) != 0x00) // values other than 0x0-0x1 are invalid
+        throw TexcHere("PROXY/2.0 error: invalid command");
+
+    char rawFamily = tok.uint8("family");
+
+    const char family = (rawFamily & 0xF0) >> 4;
+    if (family > 0x3) // values other than 0x0-0x3 are invalid
+        throw TexcHere("PROXY/2.0 error: invalid family");
+
+    const char proto = (rawFamily & 0x0F);
+    if (proto > 0x2) // values other than 0x0-0x2 are invalid
+        throw TexcHere("PROXY/2.0 error: invalid protocol type");
+
+    const uint16_t headerLen = tok.uint16("length") + mandatoryHeaderLen;
+
+    if (buf_.length() < headerLen)
+        return false; // need more bytes
+
+    // LOCAL connections do nothing with the extras
+    if (command == 0x00/* LOCAL*/)
+        return true;
+
+    union pax {
+        struct {        /* for TCP/UDP over IPv4, len = 12 */
+            struct in_addr src_addr;
+            struct in_addr dst_addr;
+            uint16_t src_port;
+            uint16_t dst_port;
+        } ipv4_addr;
+        struct {        /* for TCP/UDP over IPv6, len = 36 */
+            struct in6_addr src_addr;
+            struct in6_addr dst_addr;
+            uint16_t src_port;
+            uint16_t dst_port;
+        } ipv6_addr;
+        struct {        /* for AF_UNIX sockets, len = 216 */
+            uint8_t src_addr[108];
+            uint8_t dst_addr[108];
+        } unix_addr;
+    };
+
+    switch (family) {
+
+    case 0x1:  { // IPv4
+        pax ipu;
+        const SBuf rawAddr = tok.area(sizeof(pax::ipv4_addr), "ipv4_addr");
+        memcpy(&ipu, rawAddr.rawContent(), rawAddr.length());
+
+        dstIpAddr = ipu.ipv4_addr.dst_addr;
+        dstIpAddr.port(ntohs(ipu.ipv4_addr.dst_port));
+        srcIpAddr = ipu.ipv4_addr.src_addr;
+        srcIpAddr.port(ntohs(ipu.ipv4_addr.src_port));
+        break;
+    }
+
+    case 0x2:  { // IPv6
+        pax ipu;
+        const SBuf rawAddr = tok.area(sizeof(pax::ipv6_addr), "ipv6_addr");
+        memcpy(&ipu, rawAddr.rawContent(), rawAddr.length());
+
+        dstIpAddr = ipu.ipv6_addr.dst_addr;
+        dstIpAddr.port(ntohs(ipu.ipv6_addr.dst_port));
+        srcIpAddr = ipu.ipv6_addr.src_addr;
+        srcIpAddr.port(ntohs(ipu.ipv6_addr.src_port));
+        break;
+    }
+
+    case 0x3:  { // TODO: add support for AF_UNIX sockets.
+        tok.skip(sizeof(pax::unix_addr), "unix_addr");
+        break;
+    }
+
+    default: {
+        // the invalid family, we have checked already
+        assert(0);
+        break;
+    }
+    }
+
+    while (tok.parsed() < headerLen) { 
+        assert(!tok.atEnd());
+        if (!v2Message)
+            v2Message = new ProxyProtocol::Two::Message;
+        ProxyProtocol::Two::Tlv tlv(tok.uint8("pp2_tlv::type"));
+        const uint16_t valueLen = tok.uint16("pp2_tlv::length");
+        if (tok.parsed() + valueLen > headerLen)
+            throw TexcHere("PROXY/2.0 error: invalid pp2_tlv length");
```

Comment:
> The tokenizer would be unable to check this:

I believe this statement is incorrect. A (properly configured) tokenizer can (and, hence, should) check that the to-be-extracted area does not exceed the available (to tokenizer) content.

> the input buffer may contain some extra bytes (from the following HTTP message) 

The network I/O buffer may contain post-PROXY bytes, but the (properly configured) TLV headers tokenizer cannot. The latter should be given all TLV bytes and nothing else.

You will probably need two tokenizers to do this right: One to extract the PROXY start line followed by TLV headers (as an opaque area of known size) and a second one to parse the already extracted TLV headers. The first tokenizer should be used to learn the exact TLV headers size, but it will not tokenize individual TLV entries. The second tokenizer will.

Tokenizers are designed to be fairly cheap to create (for sub-areas) so there is no shame in using two tokenizers where a single tokenizer cannot properly track the end of being-tokenized input.

Reply:
> You will probably need two tokenizers

Done at 5dcc7a5.